Sming/Libraries/OtaUpgrade/BasicOtaUpgradeStream.h

```diff
@@ -0,0 +1,181 @@
+		DecryptionError,	  ///< Decryption failed. Probably wrong decryption key.
+		NoRomFoundError, ///< The file did not contain a ROM image suitable for the start address of the slot to upgrade.
+		RomTooLargeError,   ///< The contained ROM image does not fit into the application firmware slot.
+		DowngradeError,		///< Attempt to downgrade to older firmware version.
+		VerificationError,  ///< Signature/checksum verification failed - updated ROM not activated
+		FlashWriteError,	///< Error while writing to Flash memory.
+		RomActivationError, ///< Error while activating updated ROM slot.
+		OutOfMemoryError,   ///< Dynamic memory allocation failed
+		InternalError		///< An unexpected error occured.
+	};
+
+	ErrorCode errorCode = NoError; ///< Error code. Only relevant if `hasError()` returns `true`.
+
+	/** @brief Convert error code to string.
+	 * @see #errorCode
+	 */
+	static String errorToString(ErrorCode code);
+
+	/** @brief Process chunk of upgrade file.
+	 * @param data Pointer to chunk of data.
+	 * @param size Size of chunk pointed to by \a data in bytes.
+	 * @return If less than \a size, an error occured. Check \c #errorCode for more details.
+	 * @note Even if `write()` never returns less than \a size it is not guaranteed that the upgrade was successful.
+	 * 		 Always use `hasError()` to determine success.
+	 */
+	size_t write(const uint8_t* data, size_t size) override;
+
+	/** @brief Returns true if an error happened during the upgrade process.
+	 * @see `errorCode`
+	 */
+	bool hasError() const
+	{
+		return (state == StateError);
+	}
+
+	// overrides from IDataSourceStream
+	uint16_t readMemoryBlock(char* data, int bufSize) override
+	{
+		return 0;
+	}
+	virtual int available() override
+	{
+		return 0;
+	}
+	bool isFinished() override
+	{
+		return true;
+	}
+
+protected:
+	void setError(ErrorCode ec);
+
+private:
+	/** Representation of the ROM slot to receive the new firmware */
+	struct Slot {
+		uint8_t index;
+		uint32_t address;
+		uint32_t size;
+		bool updated = false;
+		/** Determine the parameters of the slot to receive the upgrade. */
+		Slot();
+	} slot;
+
+	// Instead of RbootOutputStream, the rboot write API is used directly because in a future extension the OTA file may contain data for multiple FLASH regions.
+	rboot_write_status rbootWriteStatus = {};
+
+	enum State {
+		StateError,
+		StateHeader,
+		StateRomHeader,
+		StateSkipRom,
+		StateWriteRom,
+		StateVerifyRoms,
+		StateRomsComplete
+	} state = StateHeader;
+
+#ifdef ENABLE_OTA_SIGNING
+	typedef OtaSignatureVerifier Verifier;
+	static const uint32_t expectedHeaderMagic = OTA_HEADER_MAGIC_SIGNED;
+#else
+	typedef OtaChecksumVerifier Verifier;
+	static const uint32_t expectedHeaderMagic = OTA_HEADER_MAGIC_NOT_SIGNED;
+#endif
+	Verifier verifier;
+
+	OTA_FileHeader fileHeader;
+
+	union {
+		OTA_RomHeader romHeader;
+		uint8_t verificationData[Verifier::NumBytes];
```

Comment:
Overlaying these items with a `union` seems un-necessary; it's complicating the code without any discernable benefit. I'd split them out.

Reply:
The intend was to save some bytes of RAM. Currently, it is probably not worth it, so I'll remove the union.

However, I have some half-baked plans to add file[system] support and maybe also the system parameters/RFCal blobs, which will add more headers. Then it may be worth reintroducing the union again.